crates/vm/vm.rs

```diff
@@ -127,32 +136,66 @@ fn run_evm(
     let tx_result = {
         let chain_spec = state.database().get_chain_config()?;
         #[allow(unused_mut)]
-        let mut evm_builder = Evm::builder()
-            .with_db(&mut state.0)
-            .with_block_env(block_env)
-            .with_tx_env(tx_env)
-            .modify_cfg_env(|cfg| cfg.chain_id = chain_spec.chain_id)
-            .with_spec_id(spec_id)
-            .with_external_context(
-                TracerEip3155::new(Box::new(std::io::stderr())).without_summary(),
-            );
-        cfg_if::cfg_if! {
-            if #[cfg(feature = "l2")] {
-                use revm::{Handler, primitives::{CancunSpec, HandlerCfg}};
-                use std::sync::Arc;
-
-                evm_builder = evm_builder.with_handler({
-                    let mut evm_handler = Handler::new(HandlerCfg::new(SpecId::LATEST));
-                    evm_handler.pre_execution.deduct_caller = Arc::new(mods::deduct_caller::<CancunSpec, _, _>);
-                    evm_handler.validation.tx_against_state = Arc::new(mods::validate_tx_against_state::<CancunSpec, _, _>);
-                    // TODO: Override `end` function. We should deposit even if we revert.
-                    // evm_handler.pre_execution.end
-                    evm_handler
-                });
+        match state {
+            EvmState::Store(db) => {
+                let mut evm_builder = Evm::builder()
+                    .with_db(db)
+                    .with_block_env(block_env)
+                    .with_tx_env(tx_env)
+                    .modify_cfg_env(|cfg| cfg.chain_id = chain_spec.chain_id)
+                    .with_spec_id(spec_id)
+                    .with_external_context(
+                        TracerEip3155::new(Box::new(std::io::stderr())).without_summary(),
+                    );
+                cfg_if::cfg_if! {
+                    if #[cfg(feature = "l2")] {
+                        use revm::{Handler, primitives::{CancunSpec, HandlerCfg}};
```

Comment:
This looks like a lot of code repetition, maybe we could separate the setup of the evm builder in two steps so we could have something like
```
let mut evm_builder = match state {
         FirstVariant => Evm::builder().with_db()
         SecondVariant => Evm::builder().with_db()
};
let mut evm_builder = evm_builder.with_x().with_y();
```
We could also directly call build on the second statement

Reply:
Yep I tried that but the problem is that `evm_builder`'s type has the DB as a generic (which can be a `Store` or `CacheDB<ExecutionDB>`) so each of the match's arms return a different type, which isn't allowed.

Reply:
Oh, makes sense, then maybe we can add a generic helper function `build_evm` that takes a `T: Database` argument and returns an `Evm<T>`?

Reply:
I missed a detail: the `Evm` and `EvmBuilder` have two and three generic types respectively, so I think such a function would be too complex for what it accomplishes.

I found another solution! the discrimination occurs at the `with_db()` call, so by moving it to the end of the chain we can minimize the duplicated code. I applied those changes in the latest commit.